File: api/insta-stalk.js
const axios = require('axios');

module.exports = async (req, res) => {
  // Enable CORS
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'GET, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type');

  if (req.method === 'OPTIONS') {
    return res.status(200).end();
  }

  if (req.method !== 'GET') {
    return res.status(405).json({ error: 'Method not allowed' });
  }

  const { username } = req.query;

  if (!username) {
    return res.status(400).json({
      search_metadata: { 
        status: "Error", 
        message: "Username required" 
      },
      profile: null
    });
  }

  const cleanUsername = username.replace('@', '');
  console.log(`🔍 Stalking: ${cleanUsername}`);

  // **MASSIVE LIST of Working APIs** (No keys needed!)
  const apis = [
    // Instagram Scrapers (Most Reliable)
    `https://instagram-scraper-api2.vercel.app/api/user-info?username=${cleanUsername}`,
    `https://igstalk.vercel.app/api/user?username=${cleanUsername}`,
    `https://insta-scraper-2024.vercel.app/api/user?username=${cleanUsername}`,
    `https://instagram-scraper-2025.vercel.app/api/info?user=${cleanUsername}`,
    
    // Indonesian API Providers (Very Reliable)
    `https://api.akuari.my.id/downloader/igstalk?username=${cleanUsername}`,
    `https://api.ryzendesu.vip/api/downloader/igdl?url=https://instagram.com/${cleanUsername}`,
    `https://rest-api.akuari.my.id/downloader/igstalk?username=${cleanUsername}`,
    `https://api.vreden.my.id/api/igstalk?username=${cleanUsername}`,
    `https://api.lolhuman.xyz/api/instagram/${cleanUsername}?apikey=beta`,
    
    // Public Proxies
    `https://corsproxy.io/?https://www.instagram.com/${cleanUsername}/?__a=1&__d=dis`,
    `https://api.codetabs.com/v1/proxy?quest=https://www.instagram.com/${cleanUsername}/?__a=1`,
    
    // HTML Scraping via Proxy (Last Resort)
    `https://api.allorigins.win/raw?url=https://www.instagram.com/${cleanUsername}/`
  ];

  // Rotating User Agents (Makes us look like real users)
  const userAgents = [
    'Mozilla/5.0 (iPhone; CPU iPhone OS 16_6 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.6 Mobile/15E148 Safari/604.1',
    'Mozilla/5.0 (Linux; Android 13; SM-S918B) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/112.0.0.0 Mobile Safari/537.36',
    'Mozilla/5.0 (iPhone; CPU iPhone OS 17_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.0 Mobile/15E148 Safari/604.1',
    'Mozilla/5.0 (Linux; Android 14; Pixel 7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Mobile Safari/537.36',
    'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.0 Safari/605.1.15',
    'Mozilla/5.0 (iPad; CPU OS 17_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.0 Mobile/15E148 Safari/604.1'
  ];

  // Try each API with rotating user agents
  for (let i = 0; i < apis.length; i++) {
    const apiUrl = apis[i];
    const randomUA = userAgents[Math.floor(Math.random() * userAgents.length)];
    
    try {
      console.log(`🔄 Trying API ${i + 1}/${apis.length}: ${apiUrl.split('/')[2]}`);
      
      const response = await axios.get(apiUrl, {
        timeout: 8000,
        headers: {
          'User-Agent': randomUA,
          'Accept': 'application/json, text/html, */*',
          'Accept-Language': 'en-US,en;q=0.9,bn;q=0.8',
          'Accept-Encoding': 'gzip, deflate, br',
          'Connection': 'keep-alive',
          'Cache-Control': 'no-cache',
          'Pragma': 'no-cache',
          'Sec-Fetch-Dest': 'document',
          'Sec-Fetch-Mode': 'navigate',
          'Sec-Fetch-Site': 'none',
          'Upgrade-Insecure-Requests': '1'
        }
      });

      // Try to extract data from various response formats
      const data = response.data;
      
      // Format 1: Instagram's own JSON
      if (data?.graphql?.user) {
        const u = data.graphql.user;
        return res.json({
          search_metadata: { status: "Success", method: "instagram-direct" },
          profile: {
            username: u.username,
            name: u.full_name,
            bio: u.biography,
            followers: u.edge_followed_by.count,
            following: u.edge_follow.count,
            posts: u.edge_owner_to_timeline_media.count,
            is_private: u.is_private,
            is_verified: u.is_verified,
            avatar: u.profile_pic_url_hd || u.profile_pic_url,
            external_link: u.external_url,
            business_category: u.business_category_name,
            business_email: u.business_email
          }
        });
      }
      
      // Format 2: Scraper API format
      if (data?.success && data?.data) {
        const u = data.data;
        return res.json({
          search_metadata: { status: "Success", method: "scraper-api" },
          profile: {
            username: u.username || u.user?.username,
            name: u.full_name || u.fullName || u.user?.full_name,
            bio: u.biography || u.bio || u.user?.biography,
            followers: u.follower_count || u.followers || u.user?.follower_count,
            following: u.following_count || u.following || u.user?.following_count,
            posts: u.media_count || u.posts || u.user?.media_count,
            is_private: u.is_private || u.private || u.user?.is_private,
            is_verified: u.is_verified || u.verified || u.user?.is_verified,
            avatar: u.profile_pic_url_hd || u.profile_pic_url || u.avatar,
            external_link: u.external_url || u.bio_links?.[0]?.url
          }
        });
      }
      
      // Format 3: akuari.my.id format
      if (data?.status && data?.data) {
        const u = data.data;
        return res.json({
          search_metadata: { status: "Success", method: "akuari" },
          profile: {
            username: u.username,
            name: u.fullName,
            bio: u.biography,
            followers: u.followers,
            following: u.following,
            posts: u.posts,
            is_private: u.private,
            is_verified: u.verified,
            avatar: u.profilePicHD || u.profilePic,
            external_link: u.externalUrl
          }
        });
      }
      
      // Format 4: HTML Scraping via Proxy
      if (typeof data === 'string' && data.includes('window._sharedData')) {
        const jsonMatch = data.match(/window\._sharedData = ({.*?});<\/script>/);
        if (jsonMatch) {
          const jsonData = JSON.parse(jsonMatch[1]);
          const user = jsonData.entry_data?.ProfilePage?.[0]?.graphql?.user;
          if (user) {
            return res.json({
              search_metadata: { status: "Success", method: "html-scrape" },
              profile: {
                username: user.username,
                name: user.full_name,
                bio: user.biography,
                followers: user.edge_followed_by.count,
                following: user.edge_follow.count,
                posts: user.edge_owner_to_timeline_media.count,
                is_private: user.is_private,
                is_verified: user.is_verified,
                avatar: user.profile_pic_url_hd,
                external_link: user.external_url
              }
            });
          }
        }
      }

    } catch (error) {
      console.log(`❌ API ${i + 1} failed: ${error.message}`);
      continue; // Try next API
    }
  }

  // ALL APIS FAILED - Last resort: Try direct HTML with different user agents
  console.log('⚠️ All APIs failed, trying direct HTML with rotating UAs...');
  
  for (const ua of userAgents) {
    try {
      const directResponse = await axios.get(`https://www.instagram.com/${cleanUsername}/`, {
        timeout: 5000,
        headers: {
          'User-Agent': ua,
          'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
          'Accept-Language': 'en-US,en;q=0.9',
          'Accept-Encoding': 'gzip, deflate, br',
          'Connection': 'keep-alive',
          'Upgrade-Insecure-Requests': '1'
        },
        maxRedirects: 5,
        validateStatus: function (status) {
          return status < 500; // Accept any status less than 500
        }
      });

      const html = directResponse.data;
      
      // Try to extract from meta tags (always works)
      const metaMatch = html.match(/<meta property="og:title" content="([^"]+)"/);
      const imageMatch = html.match(/<meta property="og:image" content="([^"]+)"/);
      const descMatch = html.match(/<meta property="og:description" content="([^"]+)"/);
      
      if (metaMatch || imageMatch) {
        const nameParts = metaMatch ? metaMatch[1].split('(')[0].trim() : cleanUsername;
        const desc = descMatch ? descMatch[1] : '';
        const followerMatch = desc.match(/([\d,]+)\s*Followers/);
        
        return res.json({
          search_metadata: { 
            status: "Success", 
            method: "meta-tags",
            note: "Limited data from meta tags"
          },
          profile: {
            username: cleanUsername,
            name: nameParts,
            bio: desc,
            followers: followerMatch ? parseInt(followerMatch[1].replace(/,/g, '')) : 'N/A',
            following: 'N/A',
            posts: 'N/A',
            is_private: html.includes('"is_private":true'),
            is_verified: html.includes('"is_verified":true'),
            avatar: imageMatch ? imageMatch[1] : null,
            external_link: null
          }
        });
      }
    } catch (e) {
      continue;
    }
  }

  // Everything failed
  res.status(404).json({
    search_metadata: {
      status: "Error",
      username: cleanUsername,
      message: "User not found or Instagram is blocking. Try these popular accounts to test: cristiano, leomessi, kimkardashian"
    },
    profile: null
  });
};

File: api/insta_stalk.py
# api/insta_stalk.py
from fastapi import FastAPI, Query, HTTPException
from fastapi.middleware.cors import CORSMiddleware
from fastapi.responses import JSONResponse
import httpx
import re
import json
from typing import Optional, Dict, Any
from pydantic import BaseModel
import random
import asyncio
from bs4 import BeautifulSoup
import time

app = FastAPI(title="Instagram Stalker API", description="Get Instagram user info by username")

# Enable CORS
app.add_middleware(
    CORSMiddleware,
    allow_origins=["*"],
    allow_methods=["GET", "OPTIONS"],
    allow_headers=["*"],
)

# Models
class Profile(BaseModel):
    username: str
    full_name: Optional[str] = None
    biography: Optional[str] = None
    followers: Optional[int] = None
    following: Optional[int] = None
    posts: Optional[int] = None
    is_private: bool = False
    is_verified: bool = False
    profile_pic_url: Optional[str] = None
    profile_pic_url_hd: Optional[str] = None
    external_url: Optional[str] = None
    business_category: Optional[str] = None
    business_email: Optional[str] = None
    business_phone: Optional[str] = None
    
class StalkResponse(BaseModel):
    status: str
    message: Optional[str] = None
    data: Optional[Profile] = None

# User agents for rotation
USER_AGENTS = [
    'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.0 Safari/605.1.15',
    'Mozilla/5.0 (iPhone; CPU iPhone OS 17_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.0 Mobile/15E148 Safari/604.1',
    'Mozilla/5.0 (Linux; Android 14; SM-S918B) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Mobile Safari/537.36',
    'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'
]

class InstagramScraper:
    def __init__(self):
        self.client = httpx.AsyncClient(
            timeout=30.0,
            follow_redirects=True,
            headers=self._get_headers()
        )
    
    def _get_headers(self) -> Dict:
        """Generate random headers"""
        return {
            'User-Agent': random.choice(USER_AGENTS),
            'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
            'Accept-Language': 'en-US,en;q=0.5',
            'Accept-Encoding': 'gzip, deflate, br',
            'DNT': '1',
            'Connection': 'keep-alive',
            'Upgrade-Insecure-Requests': '1',
            'Sec-Fetch-Dest': 'document',
            'Sec-Fetch-Mode': 'navigate',
            'Sec-Fetch-Site': 'none',
            'Cache-Control': 'max-age=0',
        }
    
    async def close(self):
        await self.client.aclose()
    
    async def extract_from_json(self, html: str) -> Optional[Profile]:
        """Extract user data from Instagram's shared data JSON"""
        try:
            # Method 1: Find window._sharedData
            shared_data_match = re.search(r'window\._sharedData\s*=\s*({.*?});</script>', html)
            if shared_data_match:
                data = json.loads(shared_data_match.group(1))
                user = data.get('entry_data', {}).get('ProfilePage', [{}])[0].get('graphql', {}).get('user')
                if user:
                    return self._parse_user_data(user)
            
            # Method 2: Find __additionalDataLoaded
            additional_data_match = re.search(r'__additionalDataLoaded\s*\([^,]+,\s*({.*?})\);', html)
            if additional_data_match:
                data = json.loads(additional_data_match.group(1))
                user = data.get('graphql', {}).get('user')
                if user:
                    return self._parse_user_data(user)
            
            # Method 3: Look for any JSON containing user data
            json_objects = re.findall(r'<script type="text\/javascript">([^<]+?)<\/script>', html)
            for script in json_objects:
                if '"user"' in script and 'edge_followed_by' in script:
                    # Try to extract JSON object
                    json_match = re.search(r'({.*"user".*})', script)
                    if json_match:
                        try:
                            data = json.loads(json_match.group(1))
                            if 'user' in data:
                                return self._parse_user_data(data['user'])
                            elif 'graphql' in data and 'user' in data['graphql']:
                                return self._parse_user_data(data['graphql']['user'])
                        except:
                            continue
        except Exception as e:
            print(f"JSON extraction error: {e}")
        return None
    
    def _parse_user_data(self, user_data: Dict) -> Profile:
        """Parse user data into Profile model"""
        # Handle different data structures
        if 'edge_followed_by' in user_data:
            # Modern Instagram format
            followers = user_data.get('edge_followed_by', {}).get('count')
            following = user_data.get('edge_follow', {}).get('count')
            posts = user_data.get('edge_owner_to_timeline_media', {}).get('count')
        else:
            # Alternative format
            followers = user_data.get('followers', 0) or user_data.get('follower_count', 0)
            following = user_data.get('following', 0) or user_data.get('following_count', 0)
            posts = user_data.get('posts', 0) or user_data.get('media_count', 0)
        
        return Profile(
            username=user_data.get('username', ''),
            full_name=user_data.get('full_name', ''),
            biography=user_data.get('biography', ''),
            followers=int(followers) if followers else 0,
            following=int(following) if following else 0,
            posts=int(posts) if posts else 0,
            is_private=user_data.get('is_private', False),
            is_verified=user_data.get('is_verified', False),
            profile_pic_url=user_data.get('profile_pic_url', ''),
            profile_pic_url_hd=user_data.get('profile_pic_url_hd', user_data.get('profile_pic_url', '')),
            external_url=user_data.get('external_url', ''),
            business_category=user_data.get('business_category_name', ''),
            business_email=user_data.get('business_email', ''),
            business_phone=user_data.get('business_phone_number', '')
        )
    
    async def extract_from_meta(self, html: str, username: str) -> Optional[Profile]:
        """Extract basic info from meta tags when JSON fails"""
        try:
            soup = BeautifulSoup(html, 'html.parser')
            
            # Get meta tags
            title_tag = soup.find('meta', property='og:title')
            image_tag = soup.find('meta', property='og:image')
            desc_tag = soup.find('meta', property='og:description')
            
            if title_tag and image_tag:
                title = title_tag.get('content', '')
                full_name = title.split('(')[0].strip() if '(' in title else title
                
                description = desc_tag.get('content', '') if desc_tag else ''
                
                # Try to extract follower count from description
                follower_match = re.search(r'([\d,]+)\s*Followers', description)
                followers = int(follower_match.group(1).replace(',', '')) if follower_match else 0
                
                return Profile(
                    username=username,
                    full_name=full_name,
                    biography=description,
                    followers=followers,
                    following=0,
                    posts=0,
                    is_private='private' in html.lower(),
                    is_verified='verified' in html.lower() or 'blue-badge' in html,
                    profile_pic_url=image_tag.get('content', ''),
                    profile_pic_url_hd=image_tag.get('content', '')
                )
        except Exception as e:
            print(f"Meta extraction error: {e}")
        return None
    
    async def stalk(self, username: str, max_retries: int = 3) -> StalkResponse:
        """Main method to stalk Instagram user"""
        clean_username = username.replace('@', '').strip()
        
        for attempt in range(max_retries):
            try:
                print(f"🔍 Attempt {attempt + 1}/{max_retries} for @{clean_username}")
                
                # Rotate user agent for each attempt
                self.client.headers.update({'User-Agent': random.choice(USER_AGENTS)})
                
                # Add delay between attempts
                if attempt > 0:
                    await asyncio.sleep(attempt * 2)
                
                # Request Instagram profile page
                response = await self.client.get(f'https://www.instagram.com/{clean_username}/')
                
                if response.status_code == 200:
                    html = response.text
                    
                    # Try JSON extraction first
                    profile = await self.extract_from_json(html)
                    if profile:
                        return StalkResponse(
                            status="success",
                            message="Profile found via JSON data",
                            data=profile
                        )
                    
                    # Fallback to meta tag extraction
                    profile = await self.extract_from_meta(html, clean_username)
                    if profile:
                        return StalkResponse(
                            status="success",
                            message="Profile found via meta tags (limited data)",
                            data=profile
                        )
                    
                    # If profile not found in HTML
                    if 'The link you followed may be broken' in html or 'Page Not Found' in html:
                        return StalkResponse(
                            status="error",
                            message=f"User '{clean_username}' does not exist"
                        )
                
                elif response.status_code == 404:
                    return StalkResponse(
                        status="error",
                        message=f"User '{clean_username}' not found"
                    )
                
                elif response.status_code == 429:
                    if attempt < max_retries - 1:
                        print(f"Rate limited, retrying after delay...")
                        continue
                    else:
                        return StalkResponse(
                            status="error",
                            message="Rate limited by Instagram. Please try again later."
                        )
                
                else:
                    print(f"Unexpected status code: {response.status_code}")
                    
            except httpx.TimeoutException:
                print(f"Timeout on attempt {attempt + 1}")
                if attempt == max_retries - 1:
                    return StalkResponse(
                        status="error",
                        message="Request timeout. Instagram might be blocking or slow."
                    )
                    
            except Exception as e:
                print(f"Error on attempt {attempt + 1}: {str(e)}")
                if attempt == max_retries - 1:
                    return StalkResponse(
                        status="error",
                        message=f"Failed to fetch profile: {str(e)}"
                    )
        
        return StalkResponse(
            status="error",
            message="All attempts failed. Instagram might be blocking automated requests."
        )

@app.get("/")
async def root():
    return {
        "name": "Instagram Stalker API",
        "version": "1.0.0",
        "description": "Get Instagram user information by username",
        "endpoints": {
            "/stalk": "GET - Get user info (use ?username=someone)",
            "/stalk/{}": "GET - Alternative path format"
        },
        "example": "/stalk?username=cristiano"
    }

@app.get("/stalk", response_model=StalkResponse)
async def stalk_user(username: str = Query(..., description="Instagram username to stalk")):
    """
    Get Instagram user information by username
    Example: /stalk?username=cristiano
    """
    scraper = InstagramScraper()
    try:
        result = await scraper.stalk(username)
        return result
    finally:
        await scraper.close()

@app.get("/stalk/{username}", response_model=StalkResponse)
async def stalk_user_path(username: str):
    """
    Alternative path format: /stalk/cristiano
    """
    scraper = InstagramScraper()
    try:
        result = await scraper.stalk(username)
        return result
    finally:
        await scraper.close()

@app.options("/stalk")
async def stalk_options():
    return JSONResponse(content={}, headers={
        "Access-Control-Allow-Origin": "*",
        "Access-Control-Allow-Methods": "GET, OPTIONS",
        "Access-Control-Allow-Headers": "*"
    })

# For Vercel serverless deployment
app = app